### PR TITLE
Add Listing and finding recipes to the experimental notebook API

### PR DESCRIPTION
**Pull request by stefsmeets**
Hi all, this code implements some new functionality for the experimental API we are building. It introduces some functions to list recipes, find recipes using some search patterns, and display nicely formatted recipe metadata in a jupyter notebook.

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Follows up on #868 
For more discussion, see issue #498 
Documentation: https://esmvaltool--901.org.readthedocs.build/projects/ESMValCore/en/901/api/esmvalcore.api.html 
Demo notebook: https://gist.github.com/stefsmeets/f4b7fc7fffe78cecb895b12bfb8ea671

### CODE REVIEW

File: esmvalcore/experimental/recipe_info.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+            Choose the renderer for the string representation.
+            Must be one of 'plaintext', 'markdown'
+        """
+        bullet = '\n - '
+        string = f'## {self.name}'
+
+        string += '\n\n'
+        string += f'{self.description}'
+
+        string += '\n\n### Contributors'
```
Comment (nielsdrost) [inline esmvalcore/experimental/recipe_info.py:202]: I think this should be authors, and mainainters is perhaps missing.

**Review by nielsdrost — COMMENTED** [review]
Nice work @stefsmeets! Works like a charm (after the bibtex fixes got merged)

There seem to be a few Codacy issues related to imports, could you have a look?

File: esmvalcore/experimental/recipe_info.py (hunk 1 above)
Comment (stefsmeets) [reply esmvalcore/experimental/recipe_info.py:202]: Yep, good point!

**Comment by stefsmeets** [comment]
> Nice work @stefsmeets! Works like a charm (after the bibtex fixes got merged)
> 
> There seem to be a few Codacy issues related to imports, could you have a look?

Yep:
- Codacy ignores the `# noqa: E402` tag, which is about the import order, because I want to have the warning at the top.
- The other one (`third party import "import yaml" should be placed before "import pybtex" (wrong-import-order)`) is odd. `isort` sorts `pybtex` before `yaml` and locally, `prospector` does not complain... So I think we can ignore this one.

**Review by nielsdrost — APPROVED** [review]
Thank @stefsmeets, looks great!

**Pull request merged**